### ISSUE TITLE
Compatibility Net-SNMP 5.7/5.8 and 5.9

### ISSUE THREAD

**Issue by Flow86** [net-snmp/net-snmp#598]
Hi,

I have a snmpd version 5.7.3 running (older suse machine).
That machine has a large physical memory (2TB)

When requesting the hrStorage table with the same snmpget  (version 5.7.3) I get the proper (non truncated) value back, even though it overflows the 32bit integer in that table.

When trying to read it with an up to date system with version 5.9.x, I only get a timeout. In the trace I get this:

```
...
trace: snmp_pdu_parse(): ..\..\snmplib\snmp_api.c, 4740:
dumph_recv:           VarBind
trace: snmp_parse_var_op(): ..\..\snmplib\snmp.c, 165:
dumph_recv:             Name
dumpx_recv:              06 0B 2B 06 01 02 01 19 02 03 01 05 01 
dumpv_recv:                ObjID: iso.3.6.1.2.1.25.2.3.1.5.1
trace: snmp_pdu_parse(): ..\..\snmplib\snmp_api.c, 4749:
dumph_recv:             Value
trace: snmp_pdu_parse(): ..\..\snmplib\snmp_api.c, 4880:
recv: error while parsing VarBindList: (parse int length 5 too large: exceeds 4)
security service 3 error parsing ScopedPDU
trace: _snmp_parse(): ..\..\snmplib\snmp_api.c, 4391:
snmp_parse: Parsed SNMPv3 message (secName:*, secLevel:*): ASN.1 parse error in message
trace: _sess_process_packet_parse_pdu(): ..\..\snmplib\snmp_api.c, 5678:
sess_process_packet: received message id#* reqid#* len 190
trace: _sess_process_packet_parse_pdu(): ..\..\snmplib\snmp_api.c, 5681:
sess_process_packet: parse fail
...
```

I assume that this is a bug in the old snmpd, but is there a way to handle it in the new version somehow?

`recv: error while parsing VarBindList: (parse int length 5 too large: exceeds 4)`

**Comment by Flow86** [comment]
I was able to reproduce this behavior with any OID:

1. Installed snmpd in a ubuntu 20.04 (version 5.8.x)
2. added to snmpd.conf:
```
view   systemonly  included   .1.3.9950.1.1
pass .1.3.9950.1.1.0 /etc/snmp/test.sh
```
* /etc/snmp/test.sh
```
#!/bin/sh -f

echo .1.3.9950.1.1.0
echo integer
echo 3508805292
```

3. Now reading the value (with snmp 5.8 inside the same ubuntu)

```
# snmpwalk -v2c -c public localhost:3161 .1.3.9950.1.1.0
iso.3.9950.1.1.0 = INTEGER: 3508805292
```

4. Wireshark:
![image](https://github.com/net-snmp/net-snmp/assets/656249/aebf7557-8381-40d8-be47-0032f139205d)

It looks like that its sending 5 bytes, the 5th byte is 0x00?

**Comment by Flow86** [comment]
After more detailed analysis, I found out this is an issue only on systems where `sizeof(long)` is only `4` / 32bit.

it seems the only workaround I found is to check for the 5th byte and treat the value as unsigned, see patch attached

[9000-fix-integer-overflow.patch](https://github.com/net-snmp/net-snmp/files/12122119/9000-fix-integer-overflow.patch)

Even though it will return a different type (ASN_COUNTER) instead now - for my use case that works at least

**Comment by bvanassche** [comment]
The above patch is as follows:
```
 snmplib/asn1.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/snmplib/asn1.c b/snmplib/asn1.c
index 07bdcc697..f4fd66b46 100644
--- a/snmplib/asn1.c
+++ b/snmplib/asn1.c
@@ -605,6 +605,17 @@ asn_parse_int(u_char * data,
         _asn_short_err(errpre, *datalength - 1, asn_length);
         return NULL;
     }
+
+#if SIZEOF_LONG == 4
+    // for very large integers we get 5 bytes integers on older net-snmp agents
+    // this will lead to a (parse int length 5 too large: exceeds 4) error,
+    // but the first byte is always zero: e.g "00 d1 24 1e ac"
+    // we have to interpret those as unsigned, where a check for a 5th byte is in.
+    if((size_t) asn_length == intsize + 1 && *bufp == 0) {
+        data[0] = ASN_COUNTER;
+        return asn_parse_unsigned_int(data, datalength, type, intp, intsize);
+    }
+#endif
 
     if ((size_t) asn_length > intsize || (int) asn_length == 0) {
         _asn_length_err(errpre, (size_t) asn_length, intsize);
```
Without the above patch, asn_parse_int() always returns an ASN_INTEGER. The above patch changes the behavior of asn_parse_int() into returning either an ASN_INTEGER or an ASN_COUNTER. I'm concerned that this might break the callers of asn_parse_int().

Wouldn't it be better to upgrade `snmpd` from version 5.7.3 to a more recent version rather than introducing changes that risk to break snmpd?

**Comment by Flow86** [comment]
>  I'm concerned that this might break the callers of asn_parse_int().

Yes the behavior changes there, and they why I wrote that "for my use case that works at least".

> Wouldn't it be better to upgrade snmpd from version 5.7.3 to a more recent version rather than introducing changes that risk to break snmpd?

well, we have a lot of systems where we cannot change that. its even with 5.8. You can even reproduce this behavior with using the `overwrite` feature in snmpd.conf. it automatically switches large integers to counter itself.

As I said, The only real way to fix this would be to use always 64bit "long"s to read stuff, so an overflow handling can tackle this properly. But that would change the whole library
